src/twisted/internet/asyncfs.py

```diff
@@ -0,0 +1,287 @@
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+# -*- test-case-name: twisted.internet.test.test_async -*-
+"""
+This module contains asynchronous filesystem access functions and the threads-based implementation
+"""
+
+from typing import Tuple, Iterable, Optional, Callable, cast, Any
+import os
+import os.path
+import threading
+
+import attr
+from zope.interface import implementer, providedBy
+
+from twisted.internet import reactor
+from twisted.python.filepath import FilePath
+from twisted.internet.interfaces import (
+    IPushProducer,
+    IReactorFileAsync,
+    FileAsyncFlags,
+    IProducer,
+    IConsumer,
+    IAsyncReader,
+    IAsyncWriter,
+    DEFAULT_CHUNK_SIZE,
+    DEFAULT_BUFFER_MAX,
+)
+from twisted.python.threadpool import ThreadPool
+from twisted.internet.threads import (
+    deferToThreadPool,
+    blockingCallFromThread,
+)
+from twisted.internet.defer import Deferred
+from twisted.logger import Logger
+
+log = Logger()
+
+
+def openAsync(
+    path: FilePath,
+    flags: FileAsyncFlags,
+    user_reactor: Any = None,
+    pool: Optional[ThreadPool] = None,
+) -> Deferred:
+    """
+    Open a file for asynchronous access
+
+    @param flags: controls opening mode
+    - B{READ} open the file for reading (all other flags imply writing)
+    - B{OVERWRITE} the file is overwritten
+    - B{TRUNCATE} the file is truncated
+    - B{CREATE} create the file if it doesn't exist
+    - B{EXCLUSIVE} create the file (if it exists already, error)and
+
+    @param user_reactor: the reactor to use. If the reactor implements
+    L{twisted.internet.interfaces.IReactorFileAsync} this is used,
+    otherwise fallback to the threaded implementation.
+
+    @param pool: the threadpool to use if threads are used. Default is the reactor's
+    own threadpool. B{WARNING:} Reactor threadpools are generally for light work, you will
+    usually need to provide a pool.
+
+    @return: a Deferred returning L{IAsyncWriter} or L{IAsyncReader}
+    """
+    if not user_reactor:
+        user_reactor = reactor
+    if IReactorFileAsync in providedBy(user_reactor):  # pragma: no cover
```

Comment:
Why not IReactorFileAsync.providedBy(user_reactor)?  Or, preferably:

```
IReactorFileAsync(user_reactor, threadedFallbackReactor).openAsync(path, flags)
```

Also, this doesn't seem like a reasonable place for a `no cover` pragma.